content/docs/reference/programming-model.md

```diff
@@ -372,15 +372,104 @@ vpc, _ := ec2.Vpc(ctx, "vpc", &VpcArgs{}, pulumi.ResourceOpt{Provider: provider}
+Every resource managed by Pulumi has a **logical name** that you specify as an argument to its constructor. For instance, the logical name of this AWS S3 Bucket is `my-bucket`:
 
-1. It is used as part of constructing the Universal Resource Name (URN) used by the Pulumi engine to track the resource across updates.
-2. Most resource providers will use it as a default prefix for constructing the cloud-provider name of the resource.
-3. Some CLI commands use the URN.
+{{< langchoose >}}
+
+```javascript
+let b = new aws.s3.Bucket("my-bucket");
+```
+
+```typescript
+let b = new aws.s3.Bucket("my-bucket");
+```
+
+```python
+b = Bucket("my-bucket")
+```
+
+```go
+bucket, _ := s3.Bucket(ctx, "my-bucket", &s3.BucketArgs{}, pulumi.ResourceOpt{})
+```
+
+This logical name is used by Pulumi to track the identity of a resource across multiple deployments of the same program and is how Pulumi knows to choose between creating new resources versus updating existing ones.
+
+The name specified during resource creation is used in two key ways:
+
+1. As a default prefix for the resource's **physical name** in the target cloud provider.
+2. To construct the Universal Resource Name (URN) used by Pulumi to track the resource across updates.
+
+#### Auto-naming {#autonaming}
+
+A resource's logical and physical names may not match. In fact, most physical resource names in Pulumi are "auto-named." As a result, even if you call your bucket with a logical name of `my-bucket`, as shown above, the physical name will typically look like this: `my-bucket-d7c2fa0`.
+
+This random suffix is added for two reasons:
+
+* It ensures that two stacks for the same project can be deployed without risk of collisions. This helps you to multi-instance your project more easily, whether that's for many development or testing instances, or even scaling to new regions. Without auto-naming, you would need to manually distinguish these resources with different physical names.
+
+* It allows Pulumi to do zero-downtime resource updates. Certain updates require replacing resources, rather than updating them in place. This is because Pulumi can create replacements first, then update existing references to them, and finally delete the old resources. If it weren't for auto-naming, Pulumi would need to do things in a very different order: namely, it would need to delete resources first, and create new instances afterwards, which is far more impactful and leads to downtime.
+
+We understand that some end users will prefer precise names. Auto-naming can be overridden by manually specifying a physical name on your resource. Most resources offer this option by way of a `name` property that may be specified in the argument object to the constructor. It just so happens that our running example of an S3 Bucket uses `bucket` instead:
```

Comment:
Very minor - but I don’t think we tend to use phrases like “we understand” in these docs - we tend to be more fact based like “For some use cases it is important to have precise names”.

Reply:
Agree, will fix!